pkg/oidc/oidc.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2020 The Lokomotive Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package oidc
+
+import (
+	"fmt"
+	"net/url"
+
+	"github.com/hashicorp/hcl/v2"
+)
+
+const (
+	defaultClientID      = "gangway"
+	defaultUsernameClaim = "email"
+	defaultGroupsClaim   = "groups"
+)
+
+// Config deals with providing OIDC related fields to the Lokomotive
+// cluster for OIDC based Authentication via Dex and Gangway.
+type Config struct {
+	IssuerURL     string `hcl:"issuer_url,optional"`
+	ClientID      string `hcl:"client_id,optional"`
+	UsernameClaim string `hcl:"username_claim,optional"`
+	GroupsClaim   string `hcl:"groups_claim,optional"`
+}
+
+// ConfigureOIDCFlags configures the Config fields after setting default values
+// and returning a list of oidc flags and error if any.
+func (c *Config) ConfigureOIDCFlags(clusterDomain string) []string {
```

Comment:
While definitely tempting to skip returning errors here, I think we should still do that here to enforce the correctness on the caller. As we do simple conversion here (from `Config` => `[]string`), which is stateless, we can make `Validate()` method unexported to minimize the exposed API.

For example, if both `IssuerURL` and `clusterDomain` strings are empty, `fmt.Sprintf("https://dex.%s", clusterDomain)` will produce bad value.

Assuming that called do validation separately from actually using the returned flags value, the calling cost is minimal:
```go
flags, _ := c.OIDC.ConfigureOIDCFlags(clusterDomain)
c.KubeAPIServerFlags = append(c.KubeAPIServerFlags, flags)
```

I would also rename the method, to better describe what it does. In my opinion `ConfigureOIDCFlags` is ambiguous, as it does not mention **what** it **Configure**s. With `oidc.Config.ToKubeAPIServerFlags()` you can read it as "Convert OIDC configuration into kube-apiserver flags".
```suggestion
func (c *Config) ToKubeAPIServerFlags(clusterDomain string) []string {
```